### DISCUSSION TITLE
ENFUGUE Web UI v0.2.5

### DISCUSSION THREAD

**Discussion by painebenjamin** [painebenjamin/app.enfugue.ai#92] [Announcements]
<div align="center">
<img src="https://github.com/painebenjamin/app.enfugue.ai/assets/57536852/1eba3d28-b52e-472e-b027-846d33534f1f" />
</div>

# New Features
<h2 id="xl-inpainting">1. Fine-Tuned SDXL Inpainting</h2>
<div align="center">
<img src="https://github.com/painebenjamin/app.enfugue.ai/assets/57536852/2aa343ee-a6cd-4934-8d6b-e9a5ad47edbb"  height="300" />
<br />
<em>A source image courtesy of Michael James Beach via <a href="https://www.pexels.com/photo/morning-breakfast-18716876/" target="_blank">pexels.com</a></em><br />
<img src="https://github.com/painebenjamin/app.enfugue.ai/assets/57536852/4ae01eaa-8f3d-4a10-845c-6372792676c4" height="300" />
<img src="https://github.com/painebenjamin/app.enfugue.ai/assets/57536852/839df549-697c-46d4-901e-84b3fb8f255b" height="300" />
<br />
<em>(Left) Outpainting the image to the left. (Right) Inpainting a refrigerator over the shelving.</em>
</div><br />

- Added support for `sdxl-1.0-inpainting-0.1` and automatic XL inpainting checkpoint merging when enabled.
- Simply use *any* Stable Diffusion XL checkpoint as your base model and use inpainting; ENFUGUE will merge the models at runtime as long as it is enabled (leave `Create Inpainting Checkpoint when Available` checked in the settings menu.)
- The merge calculation takes some time due to the size of the model. If XL pipeline caching is enabled, a cache will be created after the first calculation, which will drastically reduce the time needed to use the inpainting model for all subsequent images.

<h2 id="freeu">2. FreeU</h2>
<div align="center">
<img src="https://github.com/painebenjamin/app.enfugue.ai/assets/57536852/3bf3bb7e-03d0-4a5d-8e2c-98d61192d741" width="600" /><br /><em>Adjusting the four FreeU factors individually (first four rows), and adjusting all four factors at once (last row.)</em><br />
<img src="https://github.com/painebenjamin/app.enfugue.ai/assets/57536852/c1447ea6-1edf-459d-8489-41a385d8c399"  width="600" /><br />
<em>The same as above, using an anime model instead.</em>
</div><br />

- Added support for [FreeU](https://github.com/ChenyangSi/FreeU), developed by members of the S-Lab at Nanyang Technological University.
- This provides a new set of tweaks for modifying a model's output without any training or added processing time.
- Activate FreeU under the "Tweaks" menu for any pipeline.

<h2 id="noise">3. Noise Offset</h2>
<div align="center">
<img src="https://github.com/painebenjamin/app.enfugue.ai/assets/57536852/39e8cbf4-9942-4594-984a-a2a2eb6506cc" width="600" />
<br />
<em>A number of noise methods, blend methods and offset amounts. Click image to view full-size.</em><br />
<img src="https://github.com/painebenjamin/app.enfugue.ai/assets/57536852/d6898ab2-0a9c-4a11-9c46-88f119b64c4d" />
<br />
<em>Noise offset options under Tweaks in the sidebar.</em>
</div><br />

- Added noise offset options under Tweaks in the sidebar.
- This allows you to inject additional noise before the denoising process using 12 different noise methods and 18 different blending methods for a total of 216 new image noising methods.
- Some combinations may result in illegible or black images. It's recommended to start with a very small noise offset and increase gradually.
- All noise options are deterministic; meaning that using the same seed will always generate the same noise.

#### Noise Examples
| Blue | Brownian Fractal | Crosshatch | Default (CPU Random) | Green | Grey | Pink | Simplex | Velvet | Violet | White |
| ------ | ---------------------- | ---------------- | ------------------------------ | --------- | ------ | ------ | ----------- | -------- | -------- | ------- |
| ![blue](https://github.com/painebenjamin/app.enfugue.ai/assets/57536852/d518ce18-47bf-42a7-9a5f-10574f1a9579) | ![brownian_fractal](https://github.com/painebenjamin/app.enfugue.ai/assets/57536852/0e842af0-27ca-4149-ac7a-1c04332c40a4) | ![crosshatch](https://github.com/painebenjamin/app.enfugue.ai/assets/57536852/d6cb5237-7b37-4476-8fe8-b5f2a2d2f812) | ![default](https://github.com/painebenjamin/app.enfugue.ai/assets/57536852/19617069-92de-45b3-b322-789e965b6bf3) | ![green](https://github.com/painebenjamin/app.enfugue.ai/assets/57536852/92175bba-3747-4a85-8c8f-c2b94e833de9) | ![grey](https://github.com/painebenjamin/app.enfugue.ai/assets/57536852/c50e8e28-5361-4cd4-95d2-da092b1720ec) | ![pink](https://github.com/painebenjamin/app.enfugue.ai/assets/57536852/5acf7705-cb26-4ee3-adbc-b5b9976b29c7) | ![simplex](https://github.com/painebenjamin/app.enfugue.ai/assets/57536852/e9a57732-7b6b-4bef-b8c1-82f953c0f6a2) | ![velvet](https://github.com/painebenjamin/app.enfugue.ai/assets/57536852/96fc3024-5df1-470d-8907-48d34c53cc29) | ![violet](https://github.com/painebenjamin/app.enfugue.ai/assets/57536852/75deb929-62af-4e77-b508-7603fcd42037) | ![white](https://github.com/painebenjamin/app.enfugue.ai/assets/57536852/118a0030-9761-4226-b2ee-382f95a2a518) |

<h2 id="tiny-vae">4. Tiny Autoencoder (VAE) Preview</h2>
- Added support for MadeByOllin's [Tiny Autoencoder](https://github.com/madebyollin/taesd) for all pipelines.
- This autoencoder provides a very small memory footprint (<20 MB) with only a moderate loss in quality. For this reason, it is perfectly suited to be used for previews during the diffusion process, leaving more memory and processing power free to generate images faster.
- Input and output will still be encoded/decoded using your configured VAE (or default if not configured.)

<h2 id="clipskip">5. CLIP Skip</h2>
<div align="center">
   <img src="https://github.com/painebenjamin/app.enfugue.ai/assets/57536852/499f7b2c-c466-4cea-84bb-d68c2ae793fd" width=400 /><br />
   <img src="https://github.com/painebenjamin/app.enfugue.ai/assets/57536852/855b63d2-e928-4765-8ce0-06f513238ad2" width=400 /><br />
   <em>Examples of CLIP Skip values using a Stable Diffusion 1.5 anime model. Click each image for full resolution.</em>
</div><br />

- Added support for CLIP Skip, an input parameter to any pipeline under "Tweaks."
- This value represents the number of layers excluded from text encoding, starting from the end. In general, you should _not_ use CLIP skip, however some models were trained specifically on a reduced layer count - particularly, many anime models perform significantly better on CLIP Skip 2.
- Thanks to [Neggles](https://github.com/huggingface/diffusers/issues/3212#issuecomment-1665318394) for the example images and history.

## 6. More Detailed Pipeline Preparation Status
<div align="center">
<img src="https://github.com/painebenjamin/app.enfugue.ai/assets/57536852/2c120832-d841-4ab5-8828-7d04cc3adc8d" /><br />
<img src="https://github.com/painebenjamin/app.enfugue.ai/assets/57536852/7acfb5bc-9a88-4c26-98a2-7b2c9a1c1206" /><br />
<em>Downloading checkpoints as files or cached model weights.</em>
</div><br />

- Added more messages during the "Preparing `<Role>` Pipeline" phase.
- When any model needs to be downloaded (default checkpoints, ControlNets, VAE, and more,) the "Preparing Pipeline" message will be replaced with what it is downloading.

## 7. More Schedulers
Added the following schedulers:
- DPM Discrete Scheduler (ADPM2) Karras
- DPM-Solver++ 2M SDE (non-Karras) and DPM-Solver++ SDE Karras
    -   Not to be confused with DPM-Solver++ 2M SDE Karras
- DPM Ancestral Discrete Scheduler (KDPM2A) Karras
- Linear Multi-Step Discrete Scheduler Karras

## 8. Relative Directories in Model Picker
- Previously, only the file name of models (checkpoints, LoRA, Lycoris and Textual Inversion) would be shown, hiding any manual organization you may have done.
- Now, if you choose to organize your models in subdirectories, the relative path will be visible in the model pickers in the user interface.

## 9. Improved Gaussian Chunking
- Removed visible horizontal and vertical edges along the sides of images made using gaussian chunking.

### Example Masks
![gaussian](https://github.com/painebenjamin/app.enfugue.ai/assets/57536852/5dce9a49-51a9-403a-a93f-11b5c6d29ac6) ![gaussian-bottom](https://github.com/painebenjamin/app.enfugue.ai/assets/57536852/c42e0d48-1c42-4e6b-ae31-045fa6fe036f) ![gaussian-left](https://github.com/painebenjamin/app.enfugue.ai/assets/57536852/321040e0-ec72-46fc-9e7d-dbd7a74abe0e) ![gaussian-left-bottom](https://github.com/painebenjamin/app.enfugue.ai/assets/57536852/1201c1db-1475-4080-bcfa-258facd53232) ![gaussian-left-right](https://github.com/painebenjamin/app.enfugue.ai/assets/57536852/2c6e1d1c-b1f2-4c67-8235-1cf0fc03e656)
![gaussian-left-right-bottom](https://github.com/painebenjamin/app.enfugue.ai/assets/57536852/e03446e3-c002-45c9-884d-ff4c625de8d7) ![gaussian-left-top](https://github.com/painebenjamin/app.enfugue.ai/assets/57536852/6bb18eff-fe55-46ec-9756-aa332393cb76) ![gaussian-left-top-bottom](https://github.com/painebenjamin/app.enfugue.ai/assets/57536852/33da9fde-696d-448a-82dc-5411d002e8fc) ![gaussian-left-top-right](https://github.com/painebenjamin/app.enfugue.ai/assets/57536852/25b0ca8c-4816-4d0c-ada5-75eca037e427) ![gaussian-right](https://github.com/painebenjamin/app.enfugue.ai/assets/57536852/b9b87fd4-156e-46a5-9a18-54378e5cd01c)
![gaussian-right-bottom](https://github.com/painebenjamin/app.enfugue.ai/assets/57536852/7da575c7-2b6f-45e1-b703-06e7f85d7eee) ![gaussian-top](https://github.com/painebenjamin/app.enfugue.ai/assets/57536852/e22ec54a-84bb-4403-865b-a5698cc56a60) ![gaussian-top-bottom](https://github.com/painebenjamin/app.enfugue.ai/assets/57536852/141af6c7-9b34-44ac-bb58-9eb33cd7bce7) ![gaussian-top-right](https://github.com/painebenjamin/app.enfugue.ai/assets/57536852/f018344b-c2ea-418d-a635-304593eb53bf) ![gaussian-top-right-bottom](https://github.com/painebenjamin/app.enfugue.ai/assets/57536852/fb730e21-52dd-4976-8b8a-a8bec6f6377b)

## 10. Improved Results Browser
- Added more columns to the results browser to make it easier to copy and paste individual settings without needing to copy the entire JSON payload.

## 11. Improved CivitAI Browser
- Added the ability to middle-click or right-click download links in the CivitAI browser window.
- Also added links to view the model or the author on CivitAI.

## 12. Improved Error Messages
- Added error messages when loading LoRA, Lycoris and Textual Inversion.
- When an `AttributeError` or `KeyError` occurs, the user will be asked to ensure they are 1.5 adaptations with 1.5 models and XL adaptations with XL models.

## 13. Copy Any Tooltip to Clipboard
- Previously, regular tooltips could not be copied to clipboard, they could only be copied if they were in a table.
- Now, all tooltips can be copied. If a tooltip is visible on the screening, Holding `Ctrl` or `Cmd` and then performing a `right-click` (context menu) will copy the tooltip to the clipboard.

# Changes
- Changed the default SDXL model from `sd_xl_base_1.0.safetensors` to `sd_xl_base_1.0_fp16_vae.safetensors`.
- Changed the default value for "Inference Steps" from `40` to `20`.
- Changed the default value for "Use Chunking" from `true` to `false`.
- Changed the image options form from being hidden by default to being visible by default.
- Made it more difficult to accidentally merge nodes.

# Bug Fixes
- Fixed an issue where loading saved `.json` files would not work with some browsers.
- Fixed an issue where the inpainting model would not change when the primary model was changed and there was no explicit inpainter set.
- Fixed an issue where, during cropped inpainting, an image that was used for both inpainting and ControlNet would only get cropped for inference, and not for Control, resulting in mismatched image sizes.
- Fixed an issue where a textual inversion input field would appear blank after reloading, even though it was properly sent to the backend.
- Fixed an issue where the `Denoising Strength` slider would not appear when enabling inpainting.
- Fixed an issue where GPU-accelerated filtering would not work in Firefox.
- Fixed an issue where inpainter or refiner VAE would not be correctly set when overriding default.

**Full Changelog**: https://github.com/painebenjamin/app.enfugue.ai/compare/0.2.4...0.2.5

# How-To Guide

## Installing and Running: Portable Distributions

Select a portable distribution if you'd like to avoid having to install other programs, or want to have an isolated executable file that doesn't interfere with other environments on your system.

### Summary

| Platform | Graphics API | File(s) | CUDA Version | Torch Version |
| --------- | -------------- | ----- | --------------- | --------------- |
| MacOS | MPS | enfugue-server-0.2.5-macos-ventura-x86_64.tar.gz | N/A | 2.2.0.dev20230928 |
| Windows | CUDA          | enfugue-server-0.2.5-win-cuda-x86_64.zip.001<br />enfugue-server-0.2.5-win-cuda-x86_64.zip.002 | 12.1.1 | 2.2.0.dev20230928 |
| Windows | CUDA+TensorRT | enfugue-server-0.2.5-win-tensorrt-x86_64.zip.001<br />enfugue-server-0.2.5-win-tensorrt-x86_64.zip.002 | 11.7.1 | 1.13.1 |
| Linux | CUDA          | enfugue-server-0.2.5-manylinux-cuda-x86_64.tar.gz.0<br/>enfugue-server-0.2.5-manylinux-cuda-x86_64.tar.gz.1<br/>enfugue-server-0.2.5-manylinux-cuda-x86_64.tar.gz.2 | 12.1.1 | 2.2.0.dev20230928 |
| Linux | CUDA+TensorRT | enfugue-server-0.2.5-manylinux-tensorrt-x86_64.tar.gz.0<br/>enfugue-server-0.2.5-manylinux-tensorrt-x86_64.tar.gz.1<br/>enfugue-server-0.2.5-manylinux-tensorrt-x86_64.tar.gz.2 | 11.7.1 | 1.13.1 |

### TensorRT or CUDA?

The primary differences between the TensorRT and CUDA packages are CUDA version (11.7 vs. 12.1) and Torch version (1.13.1 vs. 2.2.0).

For general operation, Torch 2 and CUDA 12 will outperform Torch 1 and CUDA 11 for almost all operations. However, a TensorRT engine compiled in CUDA 11.7 and Torch 1.13.1 will outperform Torch 2 inference by a factor of up to 100%.

In essence,
1. If you plan to use one style very frequently (or exclusively), and have a powerful, modern Nvidia GPU, then choose TensorRT.
2. In all other cases, choose CUDA.

### Linux

After choosing TensorRT or CUDA, download the appropriate `manylinux` files here, concatenate them and extract them. A console command to do that is:

```sh
cat enfugue-server-0.2.5* | tar -xvz
```

You are now ready to run the server with:

```sh
./enfugue-server/enfugue.sh
```

Press `Ctrl+C` to exit.

### Windows
Download the `win64` files here, and extract them using a program which allows extracting from multiple archives such as [7-Zip](https://www.7-zip.org/). 

**If you are using 7-Zip, you should _not_  extract both files independently. If they are in the same directory when you unzip the first, 7-Zip will automatically unzip the second. The second file cannot be extracted on its own.**

If you are also choosing to use TensorRT, you must perform some additional steps on Windows. Follow the steps detailed [here](https://github.com/painebenjamin/app.enfugue.ai#windows-nvidia-tensorrt-support).

Locate the file `enfugue-server.exe`, and double-click it to run it. To exit, locate the icon in the bottom-right hand corner of your screen (the system tray) and right-click it, then select `Quit`.

### MacOS

Download the the `macos` file here, then double-click it to extract the package. When you run the application using the command below, your Mac will warn you of running downloaded packages, and you will have to perform an administrator override to allow it to run - you will be prompted to do this. To avoid this, you can run an included command like so:

```sh
./enfugue-server/unquarantine.sh
```

This command finds all the files in the installation and removes the `com.apple.quarantine` `xattr` from the file. This does *not* require administrator privilege. After doing this (or if you will grant the override,) run the server with:

```sh
./enfugue-server/enfugue.sh
```

*Note: while the MacOS packages are compiled on x86 machines, they are tested and designed for the new M1/M2 ARM machines thanks to Rosetta, Apple's machine code translation system.*

## Installing and Running: Conda

To install with the provided Conda environments, you need to install a version of [Conda](https://conda.org/).

After installing Conda and configuring it so it is available to your shell or command-line, download [one of the environment files](https://github.com/painebenjamin/app.enfugue.ai/tree/main/environments) depending on your platform and graphics API.

1. First, choose `windows-`, `linux-` or `macos-` based on your platform.
3. Then, choose your graphics API:
    * If you are on MacOS, you only have access to MPS.
    * If you have a powerful next-generation Nvidia GPU (3000 series and better with at least 12 GB of VRAM), use `tensorrt` for all of the capabilities of `cuda` with the added ability to compile TensorRT engines. If you do not plan on using TensorRT, select `cuda` for the most optimized build for this API.
    * If you have any other Nvidia GPU or other CUDA-compatible device, select `cuda`.
    * Additional graphics APIs (`rocm` and `directml`) are being added and will be available soon.

Finally, using the file you downloaded, create your Conda environment:

```sh
conda env create -f <downloaded_file.yml>
```

You've now installed Enfugue and all dependencies. To run it, activate the environment and then run the installed binary.

```sh
conda activate enfugue
enfugue run
```

#### Optional: DWPose Support

To install DW Pose support (a better, faster pose and face detection model), after installing Enfugue, execute the following:

```sh
mim install "mmcv>=2.0.1"
mim install "mmdet>=3.1.0"
mim install "mmpose>=1.1.0"
```

## Installing and Running: Self-Managed Environment

If you would like to manage dependencies yourself, or want to install Enfugue into an environment to share with another Stable Diffusion UI, you can install enfugue via `pip`. This is the only method available for AMD GPU's at present.

```sh
pip install enfugue
```

If you are on Linux and want TensorRT support, execute:

```sh
pip install enfugue[tensorrt]
```

If you are on Windows and want TensorRT support, follow the steps detailed [here.](https://github.com/painebenjamin/app.enfugue.ai#windows-tensorrt-support)

# Thank you!

<hr /><em>This discussion was created from the release <a href='https://github.com/painebenjamin/app.enfugue.ai/releases/tag/0.2.5'>ENFUGUE Web UI v0.2.5</a>.</em>
